pkg/app/server.go

```diff
@@ -323,10 +336,104 @@ func RunKubeStateMetrics(ctx context.Context, opts *options.Options) error {
+		// This is to ensure that the config does not throw an EOF error due to coming up empty.
+		customConfig, err := resolveCustomResourceConfig(opts)
+		if err != nil {
+			klog.Errorf("failed to update custom resource stores")
+		}
+		// Fetch the discovered factories.
+		discovererInstance.FactoryGenerator, err = customresourcestate.FromConfig(customConfig, discovererInstance)
+		if err != nil {
+			klog.Errorf("failed to update custom resource stores: %v", err)
+		}
+		// Get families for discovered factories.
+		customFactories, err := discovererInstance.FactoryGenerator()
+		if err != nil {
+			klog.Errorf("failed to update custom resource stores: %v", err)
+		}
+		// Update the list of enabled custom resources.
+		var enabledCustomResources []string
+		for _, factory := range customFactories {
+			enabledCustomResources = append(enabledCustomResources, factory.Name())
+		}
+		// Create clients for discovered factories.
+		_, discoveredCustomResourceClients, err := createKubeClient(opts.Apiserver, opts.Kubeconfig, customFactories...)
+		if err != nil {
+			klog.Errorf("failed to update custom resource stores: %v", err)
+		}
+		// Update the store builder with the new clients.
+		storeBuilder.WithCustomResourceClients(discoveredCustomResourceClients)
+		// Inject families' constructors to the existing set of stores.
+		storeBuilder.WithCustomResourceStoreFactories(customFactories...)
+		// Update the store builder with the new custom resources.
+		if err := storeBuilder.WithEnabledResources(enabledCustomResources); err != nil {
+			klog.Errorf("failed to update custom resource stores: %v", err)
+		}
+		// Configure the generation function for the custom resource stores.
+		storeBuilder.WithGenerateCustomResourceStoresFunc(storeBuilder.DefaultGenerateCustomResourceStoresFunc())
+		// Reset the flag, if there were no errors. Else, we'll try again on the next tick.
+		// Keep retrying if there were errors.
+		discovererInstance.SafeWrite(func() {
+			discovererInstance.WasUpdated = false
+		})
+		// Run the metrics handler with updated configs.
+		olderContext, olderCancel = context.WithCancel(ctx)
+		go func() {
+			// Blocks indefinitely until the unbuffered context is cancelled to serve metrics for that duration.
+			err = m.Run(olderContext)
+			if err != nil {
+				// Check if context was cancelled.
+				select {
+				case <-olderContext.Done():
+					// Context cancelled, don't really need to log this though.
+				default:
+					klog.Errorf("%s: failed to run metrics handler: %v", err)
+				}
+			}
+		}()
+	}
+	go func() {
+		for range t.C {
+			select {
+			case <-ctx.Done():
+				klog.InfoS("context cancelled")
+				t.Stop()
+				return
+			default:
+				// Check if cache has been updated.
+				shouldGenerateMetrics := false
+				discovererInstance.SafeRead(func() {
+					shouldGenerateMetrics = discovererInstance.WasUpdated
+				})
+				if shouldGenerateMetrics {
+					olderCancel()
+					generateMetrics()
+				}
```

Comment:
we should also regenerate the metrics when the custom resource config is updated.

Reply:
Wouldn't that kill everything though ([an earlier discussion on this](https://github.com/kubernetes/kube-state-metrics/pull/1851#discussion_r1094570126))? Once KSM restarts, metrics will be regenerated as well, and with the current implementation this takes negligible amount of time, comparatively.